src/util/TransitiveModel.hxx

```diff
@@ -5,94 +5,95 @@
 };
 
 /* Deduction guide for constructors */
 template <typename Data> explicit Transition(Data data)->Transition<typename Data::MyModelType>;
 
 /**
- * @details A subject model that is tightly coupled with its operations.
- *          Specifically, for each possible transition in that Model 
- *          there exists a Transition<> type to represent it.
- *          Observers to this model must update(...) with those
- *          Transition objects, but they can only be subscribed to specific
- *          events
- * @example tests/util/TransitiveModel.cxx
+ * @class     TransitiveModel
+ * @details   A Subject equivalent to a finite state machine.
+ *            Specifically, for each possible transition in that Model 
+ *            there exists a Transition<> type to represent it.
+ *            Observers to this model must update(...) with those
+ *            Transition objects, but they can only be subscribed to specific
+ *            transition types
+ * @example   tests/util/TransitiveModel.cxx
  */
 template <typename Model> class TransitiveModel : protected Model, public Subject<Transition<Model>>
 {
 protected:
   TransitiveModel() = default;
 
 public:
-  using Event = Transition<Model>;
-  using TransitionType = typename Event::TransitionType;
+  using Notification = typename Subject<Transition<Model>>::Notification;
+  using TransitionType = typename Notification::TransitionType;
 
   static_assert(std::is_enum_v<TransitionType>, "TransitionType must be an enum type");
 
-  template <typename... Args> inline void subscribe(ObserverSPtr<Event> observer, Args... filters)
+  /**
+   * @brief   Subscribe an Observer to get notified of Transitions
+   * @param   observer The Observer
+   * @param   ilist A list of TransitionTypes to subscribe to. If empty, will subscribe to all Transitions
+   */
+  void subscribe(ObserverSPtr<Notification> observer, std::initializer_list<typename Notification::TransitionType> ilist = {})
   {
     addObserver(observer);
-    (m_Filters[observer].insert(filters), ...);
+    if (ilist.size() == 0)
+      return;
+    m_Filters[observer].insert(ilist.begin(), ilist.end());
   }
 
 private:
-  using Subject<Event>::addObserver;
-  using FilterType = TreeMap<ObserverWPtr<Event>, Set<typename Event::TransitionType>, std::owner_less<ObserverWPtr<Event>>>;
+  using Subject<Notification>::addObserver;
+  using FilterType = Map<ObserverSPtr<Notification>, Set<typename Notification::TransitionType>>;
   FilterType m_Filters;
 
-  virtual inline bool mustNotify(ObserverWPtr<Event> observer, const Event &event) const noexcept final
+  bool mustNotify(ObserverSPtr<Notification> &observer, const Notification &event) const noexcept final
   {
-    if (observer.expired())
-    {
-      return false;
-    }
-    try
-    {
-      return m_Filters.at(observer).count(event.getType()) == 1;
-    }
-    catch (...)
+    auto it = m_Filters.find(observer);
+    if (it == m_Filters.end())
     {
       return true;
```

Comment:
If I understand this correctly. I would expect that we should return false in case no observer is found.

Reply:
If an observer is part of the m_Subscribers vector, and it provided no TransitionTypes, then `If empty, will subscribe to all Transitions`. Hence, no observer is being added in the map. It is impossible that ::mustNotify is called with an observer that wasn't subscribed to at least one event in the first place.

Reply:
This design has the advantage that if all your observers are subscribed to all the Transitions, then we are not using a map at all. It will not even be allocated. We are not paying for what we don't use.